Running multiple instances may result in single instance startup failure

Running a multi-instance app of uaa sometimes results in two instances running ScimGroupBootstrap at the same time, causing one of the instances failing to start with attached stacktrace. Note the IncorrectResultSizeDataAccessException happening in ScimGroupProvisioning.

`ERROR --- DispatcherServlet: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'scimGroupBootstrap' defined in ServletContext resource [/WEB-INF/spring/scim-endpoints.xml]: Invocation of init method failed; nested exception is org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size: expected 1, actual 0
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1578)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:775)
ERROR --- DispatcherServlet: 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:861)
ERROR --- DispatcherServlet: 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:541)
ERROR --- DispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.configureAndRefreshWebApplicationContext(FrameworkServlet.java:668)
ERROR --- DispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.createWebApplicationContext(FrameworkServlet.java:634)
ERROR --- DispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.createWebApplicationContext(FrameworkServlet.java:682)
ERROR --- DispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.initServletBean(FrameworkServlet.java:494)
ERROR --- DispatcherServlet: 	at org.springframework.web.servlet.HttpServletBean.init(HttpServletBean.java:136)
ERROR --- DispatcherServlet: 	at org.cloudfoundry.identity.uaa.web.RecognizeFailureDispatcherServlet.init(RecognizeFailureDispatcherServlet.java:56)
ERROR --- DispatcherServlet: 	at org.apache.catalina.core.StandardWrapper.loadServlet(StandardWrapper.java:1140)
ERROR --- DispatcherServlet: 	at org.apache.catalina.core.StandardWrapper.load(StandardWrapper.java:1027)
ERROR --- DispatcherServlet: 	at org.apache.catalina.core.StandardContext.loadOnStartup(StandardContext.java:5038)
ERROR --- DispatcherServlet: 	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5348)
ERROR --- DispatcherServlet: 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:145)
ERROR --- DispatcherServlet: 	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:729)
ERROR --- DispatcherServlet: 	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:717)
ERROR --- DispatcherServlet: 	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1834)
ERROR --- DispatcherServlet: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
ERROR --- DispatcherServlet: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
ERROR --- DispatcherServlet: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
ERROR --- DispatcherServlet: 	at java.lang.Thread.run(Thread.java:745)
ERROR --- DispatcherServlet: Caused by: org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size: expected 1, actual 0
ERROR --- DispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:239)
ERROR --- DispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:217)
ERROR --- DispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:53)
ERROR --- DispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.bootstrap.ScimGroupBootstrap.afterPropertiesSet(ScimGroupBootstrap.java:190)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1637)
ERROR --- DispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
ERROR --- DispatcherServlet: 	... 33 more
ERROR --- RecognizeFailureDispatcherServlet: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'scimGroupBootstrap' defined in ServletContext resource [/WEB-INF/spring/scim-endpoints.xml]: Invocation of init method failed; nested exception is org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size: expected 1, actual 0
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1578)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:545)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:775)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:861)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:541)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.configureAndRefreshWebApplicationContext(FrameworkServlet.java:668)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.createWebApplicationContext(FrameworkServlet.java:634)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.createWebApplicationContext(FrameworkServlet.java:682)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.initWebApplicationContext(FrameworkServlet.java:553)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.web.servlet.FrameworkServlet.initServletBean(FrameworkServlet.java:494)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.web.servlet.HttpServletBean.init(HttpServletBean.java:136)
ERROR --- RecognizeFailureDispatcherServlet: 	at javax.servlet.GenericServlet.init(GenericServlet.java:158)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.cloudfoundry.identity.uaa.web.RecognizeFailureDispatcherServlet.init(RecognizeFailureDispatcherServlet.java:56)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.StandardWrapper.initServlet(StandardWrapper.java:1227)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.StandardWrapper.loadServlet(StandardWrapper.java:1140)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.StandardWrapper.load(StandardWrapper.java:1027)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.StandardContext.loadOnStartup(StandardContext.java:5038)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5348)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:145)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:753)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:729)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:717)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1092)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1834)
ERROR --- RecognizeFailureDispatcherServlet: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
ERROR --- RecognizeFailureDispatcherServlet: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
ERROR --- RecognizeFailureDispatcherServlet: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
ERROR --- RecognizeFailureDispatcherServlet: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
ERROR --- RecognizeFailureDispatcherServlet: 	at java.lang.Thread.run(Thread.java:745)
ERROR --- RecognizeFailureDispatcherServlet: Caused by: org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size: expected 1, actual 0
ERROR --- RecognizeFailureDispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:239)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:217)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning.update(JdbcScimGroupProvisioning.java:53)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.cloudfoundry.identity.uaa.scim.bootstrap.ScimGroupBootstrap.afterPropertiesSet(ScimGroupBootstrap.java:190)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1637)
ERROR --- RecognizeFailureDispatcherServlet: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
ERROR --- RecognizeFailureDispatcherServlet: 	... 33 more
`

--- a/server/src/main/java/org/cloudfoundry/identity/uaa/scim/bootstrap/ScimGroupBootstrap.java
+++ b/server/src/main/java/org/cloudfoundry/identity/uaa/scim/bootstrap/ScimGroupBootstrap.java
@@ -28,6 +28,7 @@ import org.cloudfoundry.identity.uaa.zone.IdentityZoneHolder;
 import org.springframework.beans.factory.InitializingBean;
 import org.springframework.core.env.PropertySource;
 import org.springframework.core.io.support.ResourcePropertySource;
+import org.springframework.dao.IncorrectResultSizeDataAccessException;
 import org.springframework.util.StringUtils;
 
 import java.io.IOException;
@@ -187,7 +188,16 @@ public class ScimGroupBootstrap implements InitializingBean {
             String description = groups.get(g.getDisplayName());
             if (StringUtils.hasText(description)) {
                 g.setDescription(description);
-                groupInfos.set(i, scimGroupProvisioning.update(g.getId(), g, IdentityZoneHolder.get().getId()));
+                try{
+                    groupInfos.set(i, scimGroupProvisioning.update(g.getId(), g, IdentityZoneHolder.get().getId()));
+                } catch(IncorrectResultSizeDataAccessException e) {
+                    ScimGroup updatedGroup = getGroup(g.getDisplayName());
+                    if(updatedGroup != null && updatedGroup.getVersion() > g.getVersion()) {
+                        logger.debug("Group has already been updated by another instance, ignore error.");
+                    } else {
+                        throw e;
+                    }
+                }
             }
         }
 

--- a/server/src/test/java/org/cloudfoundry/identity/uaa/scim/bootstrap/ScimGroupBootstrapTests.java
+++ b/server/src/test/java/org/cloudfoundry/identity/uaa/scim/bootstrap/ScimGroupBootstrapTests.java
@@ -15,6 +15,7 @@ package org.cloudfoundry.identity.uaa.scim.bootstrap;
 import org.cloudfoundry.identity.uaa.resources.jdbc.JdbcPagingListFactory;
 import org.cloudfoundry.identity.uaa.scim.ScimGroup;
 import org.cloudfoundry.identity.uaa.scim.ScimGroupMember;
+import org.cloudfoundry.identity.uaa.scim.exception.ScimResourceAlreadyExistsException;
 import org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupMembershipManager;
 import org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimGroupProvisioning;
 import org.cloudfoundry.identity.uaa.scim.jdbc.JdbcScimUserProvisioning;
@@ -26,6 +27,7 @@ import org.cloudfoundry.identity.uaa.zone.IdentityZoneHolder;
 import org.junit.Before;
 import org.junit.Test;
 import org.springframework.core.env.MapPropertySource;
+import org.springframework.dao.IncorrectResultSizeDataAccessException;
 import org.springframework.jdbc.core.JdbcTemplate;
 import org.springframework.util.StringUtils;
 
@@ -38,6 +40,11 @@ import java.util.Map;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
 
 public class ScimGroupBootstrapTests extends JdbcTestBase {
 
@@ -83,6 +90,31 @@ public class ScimGroupBootstrapTests extends JdbcTestBase {
         assertNotNull(bootstrap.getGroup("org1.engg"));
         assertNotNull(bootstrap.getGroup("org1.mgr"));
         assertNotNull(bootstrap.getGroup("org1.hr"));
+    }
+    
+
+    @Test
+    public void allowsBootstrapFromOtherInstance() throws Exception {
+        //original bootstrap
+        bootstrap.setGroups(StringUtils.commaDelimitedListToSet("multiple_bootstrap_group").stream().collect(new MapCollector<>(s -> s, s -> null)));
+        bootstrap.afterPropertiesSet();
+        
+        //mock external bootstrap in between getOrCreate and update calls
+        ScimGroup multipleBootstrapGroupBefore = bootstrap.getGroup("multiple_bootstrap_group");
+        ScimGroup multipleBootstrapGroupAfter= bootstrap.getGroup("multiple_bootstrap_group");
+        multipleBootstrapGroupAfter.setVersion(multipleBootstrapGroupAfter.getVersion() + 1);
+        
+        gDB = mock(JdbcScimGroupProvisioning.class);
+        when(gDB.create(anyObject())).thenReturn(multipleBootstrapGroupBefore);
+        when(gDB.update(anyString(), anyObject())).thenThrow(new IncorrectResultSizeDataAccessException(1, 0));
+        when(gDB.query(anyString())).thenReturn(Arrays.asList(multipleBootstrapGroupAfter));
+
+        //second bootstrap
+        bootstrap = new ScimGroupBootstrap(gDB, uDB, mDB);
+        bootstrap.setGroups(StringUtils.commaDelimitedListToSet("multiple_bootstrap_group").stream().collect(new MapCollector<>(s -> s, s -> s)));
+        bootstrap.afterPropertiesSet();
+        
+        assertNotNull(bootstrap.getGroup("multiple_bootstrap_group"));
     }
 
     @Test

--- a/server/src/test/java/org/cloudfoundry/identity/uaa/scim/bootstrap/ScimGroupBootstrapTests.java
+++ b/server/src/test/java/org/cloudfoundry/identity/uaa/scim/bootstrap/ScimGroupBootstrapTests.java
@@ -105,9 +105,9 @@ public class ScimGroupBootstrapTests extends JdbcTestBase {
         multipleBootstrapGroupAfter.setVersion(multipleBootstrapGroupAfter.getVersion() + 1);
         
         gDB = mock(JdbcScimGroupProvisioning.class);
-        when(gDB.create(anyObject())).thenReturn(multipleBootstrapGroupBefore);
-        when(gDB.update(anyString(), anyObject())).thenThrow(new IncorrectResultSizeDataAccessException(1, 0));
-        when(gDB.query(anyString())).thenReturn(Arrays.asList(multipleBootstrapGroupAfter));
+        when(gDB.create(anyObject(), anyString())).thenReturn(multipleBootstrapGroupBefore);
+        when(gDB.update(anyString(), anyObject(), anyString())).thenThrow(new IncorrectResultSizeDataAccessException(1, 0));
+        when(gDB.query(anyString(), anyString())).thenReturn(Arrays.asList(multipleBootstrapGroupAfter));
 
         //second bootstrap
         bootstrap = new ScimGroupBootstrap(gDB, uDB, mDB);

